ebill_paynet/models/paynet_invoice_message.py

```diff
@@ -0,0 +1,119 @@
+# Copyright 2019 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+import os
+from datetime import datetime
+
+# Needs Jinja 2.10
+from jinja2 import Environment, select_autoescape, FileSystemLoader
+from odoo import api, fields, models
+from odoo.modules.module import get_module_root
+from zeep.exceptions import Fault
+
+from ..components.api import PayNetDWS
+
+MODULE_PATH = get_module_root(os.path.dirname(__file__))
+INVOICE_TEMPLATE_2013 = 'invoice-2013A.xml'
+INVOICE_TEMPLATE_2003 = 'invoice-2003A.xml'
+TEMPLATE_DIR = MODULE_PATH + '/messages'
+
+DOCUMENT_TYPE = {'out_invoice': 'EFD', 'out_refund': 'EGS'}
+
+jinja_env = Environment(
+    loader=FileSystemLoader(TEMPLATE_DIR),
+    autoescape=select_autoescape(['xml']),
+)
+template_2013 = jinja_env.get_template(INVOICE_TEMPLATE_2013)
+template_2003 = jinja_env.get_template(INVOICE_TEMPLATE_2003)
+
+
+class PaynetInvoiceMessage(models.Model):
+    _name = "paynet.invoice.message"
+    _description = "Paynet shipment send to service"
+
+    service_id = fields.Many2one(
+        comodel_name="paynet.service",
+        string="Paynet Service",
+        required=True,
+        ondelete="restrict",
+        readonly=True,
+    )
+    invoice_id = fields.Many2one(
+        comodel_name="account.invoice", ondelete="restrict"
+    )
+    attachment_id = fields.Many2one('ir.attachment', 'PDF')
+    state = fields.Selection(
+        selection=[
+            ('draft', 'Draft'),
+            ('sent', 'Sent'),
+            ('done', 'Done'),
+            ('reject', 'Reject'),
+            ('error', 'Error'),
+        ],
+        default='draft',
+    )
+    ic_ref = fields.Char(
+        string="IC Ref", size=14, help="Document interchange reference"
+    )
+    # Set with invoice_id.number but also with returned data from server ?
+    ref = fields.Char('Reference N°', size=35)
+    ebill_account_number = fields.Char('Paynet Id', size=20)
+    payload = fields.Text('Payload sent')
+    response = fields.Text('Response recieved')
+    shipment_id = fields.Char(size=24, help='Shipment Id on Paynet service')
+
+    def _get_ic_ref(self):
+        return 'SA%012d' % self.id
+
+    @api.multi
+    def send_to_paynet(self):
+        for message in self:
+            message._generate_payload()
```

Comment:
Nitpick: it's not obvious that some part of the code need this method to be called at first (here and in the tests). I would be explicit by returning the payload and use it in the following lines (we could keep it in the record too to keep a trace of what has been generated). But this implies to put an `self.ensure_one()` in `_generate_payload`.
A matter of taste I suppose :) (not blocking)

Reply:
Sounds like good taste :wink: so simplifying